File: src/server/services/excelService.ts
import ExcelJS from 'exceljs';

export class ExcelService {
  static async createStudentReport(
    records: Array<Record<string, any>>,
    schoolName: string,
    titleMeta: { filter_desc?: string }
  ): Promise<Buffer> {
    const wb = new ExcelJS.Workbook();
    const ws = wb.addWorksheet('Student Attendance');

    // Title Block
    ws.mergeCells('A1:I1');
    const titleCell = ws.getCell('A1');
    titleCell.value = schoolName.toUpperCase();
    titleCell.font = { name: 'Calibri', size: 16, bold: true, color: { argb: 'FFFFFFFF' } };
    titleCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF0F766E' } };
    titleCell.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(1).height = 36;

    ws.mergeCells('A2:I2');
    const subCell = ws.getCell('A2');
    const genDate = new Date().toISOString().replace('T', ' ').slice(0, 16);
    subCell.value = `STUDENT ATTENDANCE REPORT  |  ${titleMeta.filter_desc || ''}  |  Generated on: ${genDate}`;
    subCell.font = { name: 'Calibri', size: 10, color: { argb: 'FFE0F2FE' } };
    subCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF115E59' } };
    subCell.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(2).height = 22;

    if (!records || records.length === 0) {
      ws.mergeCells('A4:I5');
      const emptyCell = ws.getCell('A4');
      emptyCell.value = 'No attendance records found.';
      emptyCell.font = { name: 'Calibri', size: 14, bold: true, color: { argb: 'FF64748B' } };
      emptyCell.alignment = { horizontal: 'center', vertical: 'middle' };
      const buffer = await wb.xlsx.writeBuffer();
      return Buffer.from(buffer);
    }

    const headers = [
      'Roll Number', 'Student Name', 'Class', 'Subject',
      'Date', 'Day', 'Period', 'Status', 'Teacher'
    ];

    const startRow = 4;
    const headerRow = ws.getRow(startRow);
    headerRow.height = 24;

    headers.forEach((h, idx) => {
      const cell = headerRow.getCell(idx + 1);
      cell.value = h;
      cell.font = { name: 'Calibri', size: 11, bold: true, color: { argb: 'FFFFFFFF' } };
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF134E4A' } };
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
      cell.border = {
        top: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        bottom: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        left: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        right: { style: 'thin', color: { argb: 'FFCBD5E1' } }
      };
    });

    const summaryData: Record<string, { roll: string; total: number; present: number; absent: number }> = {};

    records.forEach((rec, idx) => {
      const rowNum = startRow + 1 + idx;
      const row = ws.getRow(rowNum);
      const studentName = rec.studentName || 'Unknown';
      const rollNo = String(rec.rollNumber || '-');
      const status = String(rec.status || 'Absent');

      let dayName = '-';
      try {
        if (rec.date) {
          const d = new Date(rec.date);
          dayName = d.toLocaleDateString('en-US', { weekday: 'long' });
        }
      } catch {
        // Ignore date parse error
      }

      const rowValues = [
        rollNo,
        studentName,
        rec.className || rec.classId || '-',
        rec.subjectName || rec.subjectId || '-',
        rec.date || '-',
        dayName,
        rec.period || '-',
        status,
        rec.teacherName || 'Teacher'
      ];

      const isZebra = idx % 2 === 1;
      const bgFillColor = isZebra ? 'FFF8FAFC' : 'FFFFFFFF';

      rowValues.forEach((val, colIdx) => {
        const cell = row.getCell(colIdx + 1);
        cell.value = val;
        cell.font = { name: 'Calibri', size: 10 };
        cell.border = {
          top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          right: { style: 'thin', color: { argb: 'FFE2E8F0' } }
        };

        if (colIdx === 7) {
          // Status column
          if (status.toLowerCase() === 'present') {
            cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFDCFCE7' } };
            cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FF166534' } };
          } else {
            cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFEE2E2' } };
            cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FF991B1B' } };
          }
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        } else {
          cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: bgFillColor } };
          cell.alignment = colIdx === 0 || colIdx === 4 || colIdx === 6 ? { horizontal: 'center', vertical: 'middle' } : { horizontal: 'left', vertical: 'middle' };
        }
      });

      // Update summary
      if (!summaryData[studentName]) {
        summaryData[studentName] = { roll: rollNo, total: 0, present: 0, absent: 0 };
      }
      summaryData[studentName].total += 1;
      if (status.toLowerCase() === 'present') {
        summaryData[studentName].present += 1;
      } else {
        summaryData[studentName].absent += 1;
      }
    });

    // Summary Section
    const sumStartRow = startRow + records.length + 3;
    ws.mergeCells(`A${sumStartRow}:F${sumStartRow}`);
    const sumTitle = ws.getCell(`A${sumStartRow}`);
    sumTitle.value = 'STUDENT-WISE ATTENDANCE SUMMARY';
    sumTitle.font = { name: 'Calibri', size: 11, bold: true, color: { argb: 'FFFFFFFF' } };
    sumTitle.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1E293B' } };
    sumTitle.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(sumStartRow).height = 24;

    const sumHeaders = ['Roll No', 'Student Name', 'Total Periods', 'Present', 'Absent', 'Attendance %'];
    const sumHdrRow = ws.getRow(sumStartRow + 1);
    sumHeaders.forEach((sh, sIdx) => {
      const cell = sumHdrRow.getCell(sIdx + 1);
      cell.value = sh;
      cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FFFFFFFF' } };
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF334155' } };
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
    });

    let sRowIdx = sumStartRow + 2;
    for (const [name, stats] of Object.entries(summaryData)) {
      const row = ws.getRow(sRowIdx);
      const pct = stats.total > 0 ? ((stats.present / stats.total) * 100).toFixed(1) + '%' : '0.0%';
      const rowVals = [stats.roll, name, stats.total, stats.present, stats.absent, pct];

      rowVals.forEach((val, vIdx) => {
        const cell = row.getCell(vIdx + 1);
        cell.value = val;
        cell.font = { name: 'Calibri', size: 10 };
        cell.alignment = vIdx === 1 ? { horizontal: 'left', vertical: 'middle' } : { horizontal: 'center', vertical: 'middle' };
        cell.border = {
          top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          right: { style: 'thin', color: { argb: 'FFE2E8F0' } }
        };
      });
      sRowIdx++;
    }

    // Auto-fit column widths
    ws.columns.forEach((column) => {
      let maxLen = 12;
      column.eachCell?.({ includeEmpty: true }, (cell) => {
        const cellVal = cell.value ? String(cell.value) : '';
        if (cellVal.length > maxLen && cellVal.length < 50) {
          maxLen = cellVal.length;
        }
      });
      column.width = maxLen + 3;
    });

    const buffer = await wb.xlsx.writeBuffer();
    return Buffer.from(buffer);
  }

  static async createTeacherReport(
    records: Array<Record<string, any>>,
    schoolName: string,
    titleMeta: { filter_desc?: string }
  ): Promise<Buffer> {
    const wb = new ExcelJS.Workbook();
    const ws = wb.addWorksheet('Teacher Attendance');

    // Title Block
    ws.mergeCells('A1:F1');
    const titleCell = ws.getCell('A1');
    titleCell.value = schoolName.toUpperCase();
    titleCell.font = { name: 'Calibri', size: 16, bold: true, color: { argb: 'FFFFFFFF' } };
    titleCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF0F766E' } };
    titleCell.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(1).height = 36;

    ws.mergeCells('A2:F2');
    const subCell = ws.getCell('A2');
    const genDate = new Date().toISOString().replace('T', ' ').slice(0, 16);
    subCell.value = `TEACHER ATTENDANCE REPORT  |  ${titleMeta.filter_desc || ''}  |  Generated on: ${genDate}`;
    subCell.font = { name: 'Calibri', size: 10, color: { argb: 'FFE0F2FE' } };
    subCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF115E59' } };
    subCell.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(2).height = 22;

    if (!records || records.length === 0) {
      ws.mergeCells('A4:F5');
      const emptyCell = ws.getCell('A4');
      emptyCell.value = 'No attendance records found.';
      emptyCell.font = { name: 'Calibri', size: 14, bold: true, color: { argb: 'FF64748B' } };
      emptyCell.alignment = { horizontal: 'center', vertical: 'middle' };
      const buffer = await wb.xlsx.writeBuffer();
      return Buffer.from(buffer);
    }

    const headers = ['Teacher UID', 'Teacher Name', 'Date', 'Day', 'Status', 'Marked By'];
    const startRow = 4;
    const headerRow = ws.getRow(startRow);
    headerRow.height = 24;

    headers.forEach((h, idx) => {
      const cell = headerRow.getCell(idx + 1);
      cell.value = h;
      cell.font = { name: 'Calibri', size: 11, bold: true, color: { argb: 'FFFFFFFF' } };
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF134E4A' } };
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
      cell.border = {
        top: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        bottom: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        left: { style: 'thin', color: { argb: 'FFCBD5E1' } },
        right: { style: 'thin', color: { argb: 'FFCBD5E1' } }
      };
    });

    const summaryData: Record<string, { total: number; present: number; absent: number }> = {};

    records.forEach((rec, idx) => {
      const rowNum = startRow + 1 + idx;
      const row = ws.getRow(rowNum);
      const teacherName = rec.teacherName || 'Unknown';
      const status = String(rec.status || 'Present');

      let dayName = '-';
      try {
        if (rec.date) {
          const d = new Date(rec.date);
          dayName = d.toLocaleDateString('en-US', { weekday: 'long' });
        }
      } catch {
        // Ignore date parse error
      }

      const rowValues = [
        rec.teacherCode || rec.teacherUid || '-',
        teacherName,
        rec.date || '-',
        dayName,
        status,
        rec.markedBy || 'Principal'
      ];

      const isZebra = idx % 2 === 1;
      const bgFillColor = isZebra ? 'FFF8FAFC' : 'FFFFFFFF';

      rowValues.forEach((val, colIdx) => {
        const cell = row.getCell(colIdx + 1);
        cell.value = val;
        cell.font = { name: 'Calibri', size: 10 };
        cell.border = {
          top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          right: { style: 'thin', color: { argb: 'FFE2E8F0' } }
        };

        if (colIdx === 4) {
          if (status.toLowerCase() === 'present') {
            cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFDCFCE7' } };
            cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FF166534' } };
          } else {
            cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFEE2E2' } };
            cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FF991B1B' } };
          }
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        } else {
          cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: bgFillColor } };
          cell.alignment = colIdx === 0 || colIdx === 2 ? { horizontal: 'center', vertical: 'middle' } : { horizontal: 'left', vertical: 'middle' };
        }
      });

      if (!summaryData[teacherName]) {
        summaryData[teacherName] = { total: 0, present: 0, absent: 0 };
      }
      summaryData[teacherName].total += 1;
      if (status.toLowerCase() === 'present') {
        summaryData[teacherName].present += 1;
      } else {
        summaryData[teacherName].absent += 1;
      }
    });

    // Summary Section
    const sumStartRow = startRow + records.length + 3;
    ws.mergeCells(`A${sumStartRow}:E${sumStartRow}`);
    const sumTitle = ws.getCell(`A${sumStartRow}`);
    sumTitle.value = 'TEACHER-WISE ATTENDANCE SUMMARY';
    sumTitle.font = { name: 'Calibri', size: 11, bold: true, color: { argb: 'FFFFFFFF' } };
    sumTitle.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1E293B' } };
    sumTitle.alignment = { horizontal: 'center', vertical: 'middle' };
    ws.getRow(sumStartRow).height = 24;

    const sumHeaders = ['Teacher Name', 'Total Days', 'Present', 'Absent', 'Attendance %'];
    const sumHdrRow = ws.getRow(sumStartRow + 1);
    sumHeaders.forEach((sh, sIdx) => {
      const cell = sumHdrRow.getCell(sIdx + 1);
      cell.value = sh;
      cell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FFFFFFFF' } };
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF334155' } };
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
    });

    let sRowIdx = sumStartRow + 2;
    for (const [name, stats] of Object.entries(summaryData)) {
      const row = ws.getRow(sRowIdx);
      const pct = stats.total > 0 ? ((stats.present / stats.total) * 100).toFixed(1) + '%' : '0.0%';
      const rowVals = [name, stats.total, stats.present, stats.absent, pct];

      rowVals.forEach((val, vIdx) => {
        const cell = row.getCell(vIdx + 1);
        cell.value = val;
        cell.font = { name: 'Calibri', size: 10 };
        cell.alignment = vIdx === 0 ? { horizontal: 'left', vertical: 'middle' } : { horizontal: 'center', vertical: 'middle' };
        cell.border = {
          top: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          bottom: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          left: { style: 'thin', color: { argb: 'FFE2E8F0' } },
          right: { style: 'thin', color: { argb: 'FFE2E8F0' } }
        };
      });
      sRowIdx++;
    }

    ws.columns.forEach((column) => {
      let maxLen = 14;
      column.eachCell?.({ includeEmpty: true }, (cell) => {
        const cellVal = cell.value ? String(cell.value) : '';
        if (cellVal.length > maxLen && cellVal.length < 50) {
          maxLen = cellVal.length;
        }
      });
      column.width = maxLen + 3;
    });

    const buffer = await wb.xlsx.writeBuffer();
    return Buffer.from(buffer);
  }
}

File: services/excel_service.py
import io
import datetime
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter

class ExcelService:
    @staticmethod
    def create_student_report(records: list, school_name: str, title_meta: dict):
        wb = openpyxl.Workbook()
        ws = wb.active
        ws.title = "Student Attendance"

        # Theme Colors
        teal_fill = PatternFill(start_color="0F766E", end_color="0F766E", fill_type="solid")
        teal_sub_fill = PatternFill(start_color="115E59", end_color="115E59", fill_type="solid")
        header_fill = PatternFill(start_color="134E4A", end_color="134E4A", fill_type="solid")
        summary_hdr_fill = PatternFill(start_color="1E293B", end_color="1E293B", fill_type="solid")
        zebra_fill = PatternFill(start_color="F8FAFC", end_color="F8FAFC", fill_type="solid")
        white_fill = PatternFill(start_color="FFFFFF", end_color="FFFFFF", fill_type="solid")
        
        present_fill = PatternFill(start_color="DCFCE7", end_color="DCFCE7", fill_type="solid")
        present_font = Font(name="Calibri", size=10, bold=True, color="166534")
        
        absent_fill = PatternFill(start_color="FEE2E2", end_color="FEE2E2", fill_type="solid")
        absent_font = Font(name="Calibri", size=10, bold=True, color="991B1B")

        white_title_font = Font(name="Calibri", size=16, bold=True, color="FFFFFF")
        white_meta_font = Font(name="Calibri", size=10, color="E0F2FE")
        header_font = Font(name="Calibri", size=11, bold=True, color="FFFFFF")
        regular_font = Font(name="Calibri", size=10)
        bold_font = Font(name="Calibri", size=10, bold=True)

        thin_border = Border(
            left=Side(style='thin', color='CBD5E1'),
            right=Side(style='thin', color='CBD5E1'),
            top=Side(style='thin', color='CBD5E1'),
            bottom=Side(style='thin', color='CBD5E1')
        )

        # Title Block
        ws.merge_cells("A1:I1")
        ws["A1"] = school_name.upper()
        ws["A1"].font = white_title_font
        ws["A1"].fill = teal_fill
        ws["A1"].alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[1].height = 36

        ws.merge_cells("A2:I2")
        meta_str = f"STUDENT ATTENDANCE REPORT  |  {title_meta.get('filter_desc', '')}  |  Generated on: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M')}"
        ws["A2"] = meta_str
        ws["A2"].font = white_meta_font
        ws["A2"].fill = teal_sub_fill
        ws["A2"].alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[2].height = 22

        if not records:
            ws.merge_cells("A4:I5")
            ws["A4"] = "No attendance records found."
            ws["A4"].font = Font(name="Calibri", size=14, bold=True, color="64748B")
            ws["A4"].alignment = Alignment(horizontal="center", vertical="center")
            buffer = io.BytesIO()
            wb.save(buffer)
            buffer.seek(0)
            return buffer

        # Detailed Table Header
        headers = [
            "Roll Number", "Student Name", "Class", "Subject", 
            "Date", "Day", "Period", "Status", "Teacher"
        ]
        start_row = 4
        for col_num, h in enumerate(headers, 1):
            cell = ws.cell(row=start_row, column=col_num)
            cell.value = h
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal="center", vertical="center")
            cell.border = thin_border
        ws.row_dimensions[start_row].height = 24

        curr_row = start_row + 1
        summary_data = {} # studentName -> {roll, total, present, absent}

        for idx, rec in enumerate(records):
            student_name = rec.get("studentName", "Unknown")
            roll_no = rec.get("rollNumber", "-")
            status = rec.get("status", "Absent")

            # Day of week from date
            date_val = rec.get("date", "")
            day_name = "-"
            try:
                dt = datetime.datetime.strptime(date_val, "%Y-%m-%d")
                day_name = dt.strftime("%A")
            except Exception:
                pass

            row_data = [
                roll_no,
                student_name,
                rec.get("className", rec.get("classId", "-")),
                rec.get("subjectName", rec.get("subjectId", "-")),
                date_val,
                day_name,
                rec.get("period", "-"),
                status,
                rec.get("teacherName", "-")
            ]

            fill_to_use = zebra_fill if idx % 2 == 1 else white_fill

            for col_num, val in enumerate(row_data, 1):
                cell = ws.cell(row=curr_row, column=col_num)
                cell.value = val
                cell.font = regular_font
                cell.fill = fill_to_use
                cell.border = thin_border
                cell.alignment = Alignment(horizontal="center" if col_num in (1, 3, 5, 6, 7) else "left", vertical="center")

                # Custom highlight on status
                if col_num == 8:
                    if status.lower() == "present":
                        cell.fill = present_fill
                        cell.font = present_font
                    else:
                        cell.fill = absent_fill
                        cell.font = absent_font

            # Accumulate summary
            if student_name not in summary_data:
                summary_data[student_name] = {"roll": roll_no, "total": 0, "present": 0, "absent": 0}
            summary_data[student_name]["total"] += 1
            if status.lower() == "present":
                summary_data[student_name]["present"] += 1
            else:
                summary_data[student_name]["absent"] += 1

            ws.row_dimensions[curr_row].height = 20
            curr_row += 1

        # Summary Section
        curr_row += 2
        ws.merge_cells(f"A{curr_row}:E{curr_row}")
        ws[f"A{curr_row}"] = "ATTENDANCE SUMMARY BY STUDENT"
        ws[f"A{curr_row}"].font = Font(name="Calibri", size=12, bold=True, color="FFFFFF")
        ws[f"A{curr_row}"].fill = summary_hdr_fill
        ws[f"A{curr_row}"].alignment = Alignment(horizontal="left", vertical="center", indent=1)
        ws.row_dimensions[curr_row].height = 24
        curr_row += 1

        sum_headers = ["Roll No", "Student Name", "Total Classes", "Present", "Absent", "Attendance %"]
        for col_num, sh in enumerate(sum_headers, 1):
            cell = ws.cell(row=curr_row, column=col_num)
            cell.value = sh
            cell.font = header_font
            cell.fill = teal_fill
            cell.border = thin_border
            cell.alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[curr_row].height = 22
        curr_row += 1

        for idx, (s_name, s_info) in enumerate(sorted(summary_data.items(), key=lambda x: str(x[1]["roll"]))):
            tot = s_info["total"]
            prs = s_info["present"]
            abs_cnt = s_info["absent"]
            pct = (prs / tot * 100) if tot > 0 else 0.0

            s_row = [
                s_info["roll"],
                s_name,
                tot,
                prs,
                abs_cnt,
                f"{pct:.1f}%"
            ]
            fill_to_use = zebra_fill if idx % 2 == 1 else white_fill
            for col_num, val in enumerate(s_row, 1):
                cell = ws.cell(row=curr_row, column=col_num)
                cell.value = val
                cell.font = bold_font if col_num in (2, 6) else regular_font
                cell.fill = fill_to_use
                cell.border = thin_border
                cell.alignment = Alignment(horizontal="center" if col_num in (1, 3, 4, 5, 6) else "left", vertical="center")
            ws.row_dimensions[curr_row].height = 20
            curr_row += 1

        # Adjust column widths
        for col in ws.columns:
            max_len = 0
            col_letter = get_column_letter(col[0].column)
            for cell in col:
                val_str = str(cell.value or "")
                if len(val_str) > max_len and "\n" not in val_str and cell.row > 2:
                    max_len = len(val_str)
            ws.column_dimensions[col_letter].width = max(max_len + 4, 14)

        buffer = io.BytesIO()
        wb.save(buffer)
        buffer.seek(0)
        return buffer

    @staticmethod
    def create_teacher_report(records: list, school_name: str, title_meta: dict):
        wb = openpyxl.Workbook()
        ws = wb.active
        ws.title = "Teacher Attendance"

        teal_fill = PatternFill(start_color="0F766E", end_color="0F766E", fill_type="solid")
        teal_sub_fill = PatternFill(start_color="115E59", end_color="115E59", fill_type="solid")
        header_fill = PatternFill(start_color="134E4A", end_color="134E4A", fill_type="solid")
        summary_hdr_fill = PatternFill(start_color="1E293B", end_color="1E293B", fill_type="solid")
        zebra_fill = PatternFill(start_color="F8FAFC", end_color="F8FAFC", fill_type="solid")
        white_fill = PatternFill(start_color="FFFFFF", end_color="FFFFFF", fill_type="solid")
        
        present_fill = PatternFill(start_color="DCFCE7", end_color="DCFCE7", fill_type="solid")
        present_font = Font(name="Calibri", size=10, bold=True, color="166534")
        absent_fill = PatternFill(start_color="FEE2E2", end_color="FEE2E2", fill_type="solid")
        absent_font = Font(name="Calibri", size=10, bold=True, color="991B1B")

        white_title_font = Font(name="Calibri", size=16, bold=True, color="FFFFFF")
        white_meta_font = Font(name="Calibri", size=10, color="E0F2FE")
        header_font = Font(name="Calibri", size=11, bold=True, color="FFFFFF")
        regular_font = Font(name="Calibri", size=10)
        bold_font = Font(name="Calibri", size=10, bold=True)

        thin_border = Border(
            left=Side(style='thin', color='CBD5E1'),
            right=Side(style='thin', color='CBD5E1'),
            top=Side(style='thin', color='CBD5E1'),
            bottom=Side(style='thin', color='CBD5E1')
        )

        ws.merge_cells("A1:E1")
        ws["A1"] = school_name.upper()
        ws["A1"].font = white_title_font
        ws["A1"].fill = teal_fill
        ws["A1"].alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[1].height = 36

        ws.merge_cells("A2:E2")
        meta_str = f"TEACHER ATTENDANCE REPORT  |  {title_meta.get('filter_desc', '')}  |  Generated on: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M')}"
        ws["A2"] = meta_str
        ws["A2"].font = white_meta_font
        ws["A2"].fill = teal_sub_fill
        ws["A2"].alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[2].height = 22

        if not records:
            ws.merge_cells("A4:E5")
            ws["A4"] = "No attendance records found."
            ws["A4"].font = Font(name="Calibri", size=14, bold=True, color="64748B")
            ws["A4"].alignment = Alignment(horizontal="center", vertical="center")
            buffer = io.BytesIO()
            wb.save(buffer)
            buffer.seek(0)
            return buffer

        headers = ["Teacher UID", "Teacher Name", "Date", "Day", "Status"]
        start_row = 4
        for col_num, h in enumerate(headers, 1):
            cell = ws.cell(row=start_row, column=col_num)
            cell.value = h
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal="center", vertical="center")
            cell.border = thin_border
        ws.row_dimensions[start_row].height = 24

        curr_row = start_row + 1
        summary_data = {}

        for idx, rec in enumerate(records):
            t_uid = rec.get("teacherCode") or rec.get("teacherUid", "-")
            t_name = rec.get("teacherName", "Unknown")
            status = rec.get("status", "Absent")
            date_val = rec.get("date", "")
            day_name = "-"
            try:
                dt = datetime.datetime.strptime(date_val, "%Y-%m-%d")
                day_name = dt.strftime("%A")
            except Exception:
                pass

            row_data = [t_uid, t_name, date_val, day_name, status]
            fill_to_use = zebra_fill if idx % 2 == 1 else white_fill

            for col_num, val in enumerate(row_data, 1):
                cell = ws.cell(row=curr_row, column=col_num)
                cell.value = val
                cell.font = regular_font
                cell.fill = fill_to_use
                cell.border = thin_border
                cell.alignment = Alignment(horizontal="center" if col_num in (1, 3, 4) else "left", vertical="center")
                if col_num == 5:
                    if status.lower() == "present":
                        cell.fill = present_fill
                        cell.font = present_font
                    else:
                        cell.fill = absent_fill
                        cell.font = absent_font

            if t_name not in summary_data:
                summary_data[t_name] = {"uid": t_uid, "total": 0, "present": 0, "absent": 0}
            summary_data[t_name]["total"] += 1
            if status.lower() == "present":
                summary_data[t_name]["present"] += 1
            else:
                summary_data[t_name]["absent"] += 1

            ws.row_dimensions[curr_row].height = 20
            curr_row += 1

        # Summary Section
        curr_row += 2
        ws.merge_cells(f"A{curr_row}:E{curr_row}")
        ws[f"A{curr_row}"] = "TEACHER ATTENDANCE SUMMARY"
        ws[f"A{curr_row}"].font = Font(name="Calibri", size=12, bold=True, color="FFFFFF")
        ws[f"A{curr_row}"].fill = summary_hdr_fill
        ws[f"A{curr_row}"].alignment = Alignment(horizontal="left", vertical="center", indent=1)
        ws.row_dimensions[curr_row].height = 24
        curr_row += 1

        sum_headers = ["Teacher UID", "Teacher Name", "Total Marked Days", "Present", "Absent", "Attendance %"]
        for col_num, sh in enumerate(sum_headers, 1):
            cell = ws.cell(row=curr_row, column=col_num)
            cell.value = sh
            cell.font = header_font
            cell.fill = teal_fill
            cell.border = thin_border
            cell.alignment = Alignment(horizontal="center", vertical="center")
        ws.row_dimensions[curr_row].height = 22
        curr_row += 1

        for idx, (t_name, s_info) in enumerate(sorted(summary_data.items())):
            tot = s_info["total"]
            prs = s_info["present"]
            abs_cnt = s_info["absent"]
            pct = (prs / tot * 100) if tot > 0 else 0.0

            s_row = [
                s_info["uid"],
                t_name,
                tot,
                prs,
                abs_cnt,
                f"{pct:.1f}%"
            ]
            fill_to_use = zebra_fill if idx % 2 == 1 else white_fill
            for col_num, val in enumerate(s_row, 1):
                cell = ws.cell(row=curr_row, column=col_num)
                cell.value = val
                cell.font = bold_font if col_num in (2, 6) else regular_font
                cell.fill = fill_to_use
                cell.border = thin_border
                cell.alignment = Alignment(horizontal="center" if col_num in (1, 3, 4, 5, 6) else "left", vertical="center")
            ws.row_dimensions[curr_row].height = 20
            curr_row += 1

        for col in ws.columns:
            max_len = 0
            col_letter = get_column_letter(col[0].column)
            for cell in col:
                val_str = str(cell.value or "")
                if len(val_str) > max_len and "\n" not in val_str and cell.row > 2:
                    max_len = len(val_str)
            ws.column_dimensions[col_letter].width = max(max_len + 4, 15)

        buffer = io.BytesIO()
        wb.save(buffer)
        buffer.seek(0)
        return buffer
